### PR TITLE
vmware: Add docs for filters

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

Add a scenario guide for filters in VMware documentation

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
docs/docsite/rst/scenario_guides/guide_vmware.rst
docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst
docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_hostnames.rst

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @jainnikhil30 Could you please review this? Thanks.

**Comment by jainnikhil30** [comment]
@Akasurde LGTM

Thanks for adding this to the docs. This will be very useful info for future reference.

**Review by jainnikhil30 — APPROVED** [review]

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 1]
```diff
@@ -0,0 +1,186 @@
+.. _vmware_ansible_inventory_using_filters:
+
+*************************************************
+Using VMware dynamic inventory plugin - Filters
+*************************************************
+
+.. contents:: Topics
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
.. contents::
  :local:
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+VMware dynamic inventory plugin - Filtering VMware Guests
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
VMware dynamic inventory plugin - filtering VMware guests
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+=========================================================
+
+
+VMware inventory plugin allows you to filters VMware guests using the ``filters`` configuration parameter.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
VMware inventory plugin allows you to filter VMware guests using the ``filters`` configuration parameter.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+In this scenario guide, we will see how you configure ``filters`` for the given VMware guest in the inventory.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
This section shows how you configure ``filters`` for the given VMware guest in the inventory.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+To include tag-related information for the virtual machines in your dynamic inventory, you also need the `vSphere Automation SDK <https://code.vmware.com/web/sdk/65/vsphere-automation-python>`_, which supports REST API features such as tagging and content libraries, on your control node.
+You can install the ``vSphere Automation SDK`` following `these instructions <https://github.com/vmware/vsphere-automation-sdk-python#installing-required-python-packages>`_.
+
+.. code-block:: bash
+
+    $ pip install pyvmomi
+
+Starting in Ansible 2.10, the VMware dynamic inventory plugin is available in the ``community.vmware`` collection included Ansible.
+To install the latest ``community.vmware`` collection:
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
Alternately, to install the latest ``community.vmware`` collection:
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+.. code-block:: ini
+
+  [inventory]
+  enable_plugins = community.vmware.vmware_vm_inventory
+
+2. Create a file that ends in ``vmware.yml`` or ``vmware.yaml`` in your working directory.
+
+The ``vmware_vm_inventory`` inventory plugin takes in the same authentication information as any other VMware modules does.
+
+Let us assume we want to list all RHEL7 vms with the power state as "poweredOn", a valid inventory file with filters for the given VMware guest looks like:
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
Let us assume we want to list all RHEL7 VMs with the power state as "poweredOn". A valid inventory file with filters for the given VMware guest looks as follows:
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    - config.guestId == "rhel7_64Guest"
+    - summary.runtime.powerState == "poweredOn"
+
+
+Here, we have configured two filters -
+
+* ``config.guestId`` is equal to ``rhel7_64Guest``
+* ``summary.runtime.powerState`` is equal to ``poweredOn``
+
+This will retrieve all the virtual machine which satisfies these two conditions and populate them in the inventory.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
This retrieves all the VMs that satisfy these two conditions and populates them in the inventory.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+Notice that the conditions are combined using an ``and`` operation.
+
+Using ``or`` conditions in filters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Let us assume you want filter RHEL7 and Ubuntu vms, you can use multiple filters using ``or`` condition in your inventory file.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
Let us assume you want filter RHEL7 and Ubuntu VMs. You can use multiple filters using ``or`` condition in your inventory file.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+A valid filter in VMware inventory file will be:
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
A valid filter in the VMware inventory file for this example is:
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    with_tags: False
+    hostnames:
+    - config.name
+    filters:
+    - config.guestId == "rhel7_64Guest" or config.guestId == "ubuntu64Guest"
+
+
+You can check all allowed properties for filters for the given virtual machine at :ref:`vmware_inventory_vm_attributes`.
+
+If you are using ``properties`` parameter with custom vm properties, make sure that you include all the properties used by filters as well in your VM property list.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
If you are using the ``properties`` parameter with custom VM properties, make sure that you include all the properties used by filters as well in your VM property list.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+For example, if we want all RHEL7 and Ubuntu vms which are poweredOn, you can use inventory file:
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
For example, if we want all RHEL7 and Ubuntu vms that are poweredOn, you can use inventory file:
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    properties:
+    - 'config.name'
+    - 'config.guestId'
+    - 'guest.ipAddress'
+    - 'summary.runtime.powerState'
+    filters:
+    - config.guestId == "rhel7_64Guest" or config.guestId == "ubuntu64Guest"
+    - summary.runtime.powerState == "poweredOn"
+
+Here, we are using minimum vm properties, that is ``config.name``, ``config.guestId``, ``summary.runtime.powerState``, and ``guest.ipAddress``.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
Here, we are using minimum VM properties, that is ``config.name``, ``config.guestId``, ``summary.runtime.powerState``, and ``guest.ipAddress``.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+* ``config.name`` is used by ``hostnames`` parameter.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
* ``config.name`` is used by the ``hostnames`` parameter.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+* ``config.guestId`` and ``summary.runtime.powerState`` are used by ``filters`` parameter.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
* ``config.guestId`` and ``summary.runtime.powerState`` are used by the ``filters`` parameter.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+* ``guest.guestId`` is used by ``ansible_host`` internally by inventory plugin.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
* ``guest.guestId`` is used by ``ansible_host`` internally by the  inventory plugin.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+
+Executing ``ansible-inventory --list -i <filename>.vmware.yml`` will create a list of the virtual machines that are ready to be configured using Ansible.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
Executing ``ansible-inventory --list -i <filename>.vmware.yml`` creates a list of the virtual machines that are ready to be configured using Ansible.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+What to expect
+--------------
+
+You will notice that the inventory hosts are filtered depending upon your ``filters`` section.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
You will notice that the inventory hosts are filtered depending on your ``filters`` section.
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+                "config.name": "vm_8046",
+                "config.guestId": "rhel7_64Guest",
+                ...
+                "guest.toolsStatus": "toolsNotInstalled",
+                "summary.runtime.powerState": "poweredOn",
+            },
+        ...
+    }
+
+Troubleshooting
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
Troubleshooting VMware filters
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+---------------
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
-----------------------------------
```

File: docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+
+If the custom property specified in ``filters`` fails:
+
+- Check if the values provided for username and password are correct.
+- Make sure it is a valid property, see :ref:`vmware_inventory_vm_attributes`.
+- Use ``strict: True`` to get more information about the error.
+- Please make sure that you are using latest version VMware collection.
```
Comment (samccann) [inline docs/docsite/rst/scenario_guides/vmware_scenarios/vmware_inventory_filters.rst]:
```suggestion
- Please make sure that you are using latest version of the VMware collection.
```

**Review by samccann — COMMENTED** [review]
bonus points for having a troubleshooting section!!  :-)

**Review by samccann — APPROVED** [review]

**Pull request merged**